low-level/sensors/impl/centos_7/service_monitor.py

```diff
@@ -0,0 +1,411 @@
+
+            # Retrieve the main loop which will be called in the run method
+            self._loop = gobject.MainLoop()
+
+            # Initialize the gobject threads and get its context
+            gobject.threads_init()
+            context = self._loop.get_context()
+
+            delay = 0
+            while self.is_running():
+                # At interval of 'thread_sleep' check for events occured for
+                # registered services and process them(call on_pro_changed())
+                context.iteration(False)
+                time.sleep(self.thread_sleep)
+
+                # At interval of 'polling_freqency' process unregistered
+                # services and services with not-active state. 
+                if delay == self.polling_frequency:
+                    delay = 0
+
+                    temp_ser_to_mon = self.services_to_monitor.copy()
+                    for service in temp_ser_to_mon:
+                        self.connect_to_prop_changed_signal(service)
+
+                    self.check_notactive_services()
+
+                delay+=1
+
+            logger.debug("ServiceMonitor gracefully breaking out " +\
+                                "of dbus Loop, not restarting.")
+        except Exception as e:
+            if self.is_running:
+                logger.debug("Ungracefully breaking out of " +\
+                                "dbus loop with error: %s" % e)
+
+    def connect_to_prop_changed_signal(self, service):
+        """Connect the service to the 'PropertiesChanged' Signal."""
+        try:
+            unit = self._bus.get_object('org.freedesktop.systemd1',\
+                                        self._manager.LoadUnit(service))
+            Iunit = Interface(unit, dbus_interface='org.freedesktop.DBus.Properties')
+            state = str(Iunit.Get('org.freedesktop.systemd1.Unit', 'ActiveState'))
+            substate = str(Iunit.Get('org.freedesktop.systemd1.Unit', 'SubState'))
+            pid = str(Iunit.Get('org.freedesktop.systemd1.Service', 'ExecMainPID'))
+
+            self.update_status(service, state, substate, pid)
+
+            Iunit2 = Interface(unit,
+                            dbus_interface='org.freedesktop.systemd1.Manager')
+
+            Iunit2.connect_to_signal('PropertiesChanged',
+                                    lambda a, b, c, p = unit :
+                                    self.on_prop_changed(a, b, c, p),
+                                    dbus_interface=PROPERTIES_IFACE)
+
+            logger.debug(f"{service}({pid}) state is {state}:{substate}")
+
+            if state != "active":
+                self.failed_services.append(service)
+                self.raise_alert(service, "N/A", state, "N/A", substate,
+                                    "N/A", pid, 0)
+                logger.error(f"{service} is not active initially. state = {state}:{substate}")
+
+            self.services_to_monitor.remove(service)
+
```

Comment:
>On bootup, SSPL will scan this static set of services and create list of enabled services out of it and then monitor those services.

This requirement exists in jira description, I dont see any code to filter out disabled services out of the static list of monitored services list from sspl.conf and then monitor only enabled services.

Also, if a service is enabled but in inactive state, then it should be monitored and an immediate alert should be raised on detecting its inactive state.

Reply:
This functionality is already taken care of..